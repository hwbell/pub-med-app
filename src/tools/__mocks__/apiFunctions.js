const fetch = require('node-fetch')

// query and return based on date / citations
export async function getArticles(query, sortParam) {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  // let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;

  // let headers = {}
  // headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let list = [];
  for (let i = 0; i < 50; i++) {
    list.push({
      title: `Article ${i + 1}`,
      pmid: `PMID10000${i + 1}`
    })
  }

  let articles = await Promise.resolve({
    resultList: {
      result: list
    }
  })

  return articles;
}

// parses the result of a search query to a simple array of the article titles
export function parseSearchToTitlesArray(searchResults) {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })

  return titlesArray;
}