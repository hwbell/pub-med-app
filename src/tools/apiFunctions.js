const fetch = require('node-fetch')

// query and return based on date / citations
const getArticles = async (query, sortParam) => {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;
  
  let headers = {}
  headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let articles = await fetch(searchUrl, {
    method: 'POST',
    headers
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return articles;
}

// parses the result of a search query to a simple array of the article titles
const parseSearchToTitlesArray = (searchResults) => {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })
  return titlesArray;
}

// let results = getArticles('cancer').then((response) => {
//   console.log(response.resultList.result.slice(0,2))
// })

module.exports = {
  getArticles,
  parseSearchToTitlesArray,
}
