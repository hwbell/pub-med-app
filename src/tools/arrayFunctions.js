// this function will remove an article from a specific collection in the collections array
const removeArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles = collection.articles.filter(item => item.id !== article.id)
    }

  });

  return collections;
}

// this function will add an article to a specific collection
const addArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles.push(article);
    }

  });

  return collections;
}

module.exports = {
  removeArticle,
  addArticle
}