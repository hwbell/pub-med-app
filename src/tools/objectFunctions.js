
// this comibines two objects for the keys specified, giving priority to obj1
const combineObjects = (obj1, obj2, keys) => {

  if(!obj1 || !obj2) {
    return console.log('object argument missing')
  }
  if(!keys || !keys.length) {
    return console.log('keys argument missing')
  }

  let combined = {};

  keys.forEach((key) => {
    if (obj1[key]) {
      combined[key] = obj1[key];
    } else if (obj2[key]){
      combined[key] = obj2[key];
    } else {
      combined[key] = '';
    }
  })

  return combined;
}

module.exports = {
  combineObjects
}