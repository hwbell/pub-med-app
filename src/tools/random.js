let list = [];
for (let i = 0; i < 50; i++) {
  list.push({
    title: `Article ${i + 1}`,
    pmid: `PMID10000${i}`
  })
}

console.log(list)