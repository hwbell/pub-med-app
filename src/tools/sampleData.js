
// set some sample data for tests

export function getSampleCollections() {
  let collections = [];
  for (let i = 0; i < 10; i++) {
    collections.push({
      name: `collection ${i + 1}`,
      articles: [
        {
          name: `article ${i + 1}`,
          name: `article ${i + 2}`,
          name: `article ${i + 3}`,
        }
      ]
    })
  }
  return collections;
}