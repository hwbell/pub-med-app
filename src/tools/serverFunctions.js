const collectionServerUrl = process.env.REACT_APP_COLLECTION_SERVER_URL || 'http://localhost:3000/';
const fetch = require('node-fetch');

// this function will actually make a new user OR sign in an existing user, depending on the
// context
export async function signInUser(user, isNewUser) {
  if (!user) {
    return;
  }

  let headers = {}
  headers['content-type'] = 'application/application/json';

  let url = isNewUser ? `${collectionServerUrl}users` : `${collectionServerUrl}users/login`;
  let response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return response;
}

// patches the user's profile
export async function patchUser(patch, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}users/me`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// saves a collection to the server, or patches an exisiting collection on the server
export async function saveCollection(collection, headers, isExisting) {

  // console.log(collection)
  // assign the request params accoring to either PATCH for existing collections or 
  // POST for new collections

  // base url
  let url = `${collectionServerUrl}collections`;
  let method, body;

  if (isExisting) {
    // if its a patch, we only want to send allowed properties.
    let { name, articles } = collection;
    body = {
      name,
      articles
    }
    method = 'PATCH';
    url += `/${collection._id}`;
  } else {
    body = collection;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// deletes user's collection 
export async function deleteCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection._id}`, {
    method: 'DELETE',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// gets a user's profile / collections / threads
export async function getUserCollections(headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/me`, {
    method: 'GET',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// patches an existing collection
export async function patchCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection.id}`, {
    method: 'PATCH',
    body: JSON.stringify(collection),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// get the threads on the server, 10 at a time, according to the sortBy param
export async function getPublicThreads(headers, sortBy, page) {

  if (!sortBy || !page) {
    return console.log('Missing sortBy or page parameter.');
  }
  // page represents results 1-10, 11-20, 21-30, etc. 
  let url = `${collectionServerUrl}threads/all/${sortBy}/${page}`;
  let method = 'GET';

  let serverResponse = await fetch(url, {
    method,
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}



// saves a collection to the server, or patches an exisiting collection on the server
export async function saveThread(thread, headers, isComment) {

  // assign the request params accoring to either PATCH for existing thread or 
  // POST for new thread

  // base url
  let url = `${collectionServerUrl}threads`;
  let method, body;

  if (isComment) {
    // if its a patch, we only want to send allowed properties.
    body = {
      user: thread.user,
      text: thread.text,
      add: thread.add
    };
    method = 'PATCH';
    url += `/comments/${thread._id}`;

  } else {
    body = thread;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}








// let user = {
//   "name": makeRandomString(6),
//   "password": makeRandomString(8),
//   "email": `${makeRandomString(6)}@example.com`,
//   "age": 25
// }
// let collection = {
//   "name": "articles about science",
//   "articles": [
//     {
//       "url": "http://nature/bio/moggetarticle1"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle2"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle3"
//     }
//   ]
// }

// let signInAttempt = signInUser(user).then((response) => {

//   console.log(localStorage.token)
// })

// let headers = {
//   Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZDFhN2VlMTkzYmI0ZTQyZWM1Mjk5MmQiLCJpYXQiOjE1NjIwMTc1MDV9.AT5W-Vu8LssYs3fmH3IpnWSbQFxS0zk8S2WdW-2OAcs',
//   Accept: 'application/json',
//   'Content-Type': 'application/json',
// }
// let saveCollectionAttempt = saveCollection(collection, headers).then((response) => {
//   // 
//   console.log(response);
// })


// module.exports = {
//   signInUser,
//   saveCollection
// }