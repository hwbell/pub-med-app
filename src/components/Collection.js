import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import { Button } from 'reactstrap';
import ArticleResult from './ArticleResult';
import { PDFViewer } from '@react-pdf/renderer';
import GeneratedPdf from './GeneratedPdf';

// tools
import { saveCollection } from '../tools/serverFunctions';
import { CSSTransitionGroup } from 'react-transition-group' // ES6

// ******************************************************************************
class Collection extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      editing: false,
      showPreview: false
    }

    this.togglePreview = this.togglePreview.bind(this);
    this.postCollection = this.postCollection.bind(this);
    this.toggleEdit = this.toggleEdit.bind(this);
  }

  toggleEdit() {
    this.setState({
      editing: !this.state.editing
    });
  }

  togglePreview() {
    this.setState({
      showPreview: !this.state.showPreview
    })
  }

  postCollection() {
    console.log(`posting ${this.props.collection.name} to server`)
    let token = JSON.parse(localStorage.getItem('token'));
    let headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }

    // Post the collection to save it to the server. Using the currently stored token
    // will assign the user as the owner of the collection in mongodb on the backend
    return saveCollection(this.props.collection, headers)
      .then((response) => {
        console.log(response)

        // once we have successfully posted, we will: 

        // 1. Refresh the userCollections - this will send the new collection list sent back
        // from the server to the root App component to update all concerned components
        this.props.refreshUserCollections(response);

        // 2. Remove this collection from the 'New Collections' list 
        this.props.handleDelete(this.props.collection);

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  renderResults(collection) {

    const collectionButtons = [
      {
        text: 'view article',
        onClick: this.props.viewArticle
      },
      {
        text: 'remove',
        onClick: this.props.handleSubmit
      }
    ];

    return collection.articles.map((article, i) => {
      return (
        <ArticleResult key={i}
          collection={collection}
          article={article}
          buttons={collectionButtons} />
      )
    })

  }

  render() {
    let collection = this.props.collection;

    return (
      <div className="outline collection" style={styles.content}>
        <div className="row" style={styles.titleHolder}>

          <div className="row">
            <p className="collection-title">
              <strong>{`${collection.name} `}</strong>
            </p>
            <i className="far fa-edit" style={styles.icon}
              onClick={this.toggleEdit}></i>
            <p className="collection-title">
              {`${collection.articles.lenth}`}
            </p>
          </div>

          {/* preview / download options for the article */}
          <div style={styles.buttonHolder}>
            <Button
              className="add article-button" size="sm"
              onClick={this.togglePreview}>
              {!this.state.showPreview ? 'make pdf' : 'hide pdf'}
            </Button>
            {!this.props.isSaved &&
              <Button
                className="add article-button" size="sm"
                onClick={this.postCollection}>
                save to my collections
            </Button>}
            <Button
              className="warn article-button" size="sm"
              onClick={() => this.props.handleDelete(collection)}>
              delete
            </Button>

          </div>
        </div>

        <CSSTransitionGroup
          transitionName="fade"
          transitionEnterTimeout={500}
          transitionLeaveTimeout={300}>

          {this.state.showPreview &&
            <div ref={this.props.ref} key="pdf" className="pdf-holder">
              <PDFViewer className="pdf-viewer">
                <GeneratedPdf collection={collection} />
              </PDFViewer>
            </div>}

          {!this.state.showPreview &&
            <div ref={this.props.ref} key="results" className="results-holder">
              {this.renderResults(collection)}
            </div>}

        </CSSTransitionGroup>

      </div>

    );
  }
}

const styles = {
  buttonHolder: {
    padding: '10px'
  },
  icon: {
    marginTop: '14px',
    fontSize: '14px',
  }
}

export default Collection;
