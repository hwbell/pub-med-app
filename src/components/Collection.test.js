import React from 'react';
import Collection from './Collection';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { getArticles } from '../tools/apiFunctions';

describe('Collection', () => {

  let someProps;
  beforeAll(async () => {
    let articles = await getArticles('medicine').then((response) => {
      return response.resultList.result;
    });

    someProps = {
      collection: {
        name: 'sample collection',
        articles
      }
    }
  });

  // tests
  it('renders without crashing', async () => {
    let wrapper = shallow(<Collection {...someProps} />);
    // wrapper.update();

  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<Collection {...someProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {
    let wrapper = shallow(<Collection {...someProps} />);

    expect(wrapper.find('.collection').length).toEqual(1);
    expect(wrapper.find('.outline').length).toEqual(1);
    expect(wrapper.find('Button').length).toEqual(2);


  });

  // it('should post a collection to the server', () => {
  //   let wrapper = shallow(<Collection {...someProps} />);    
    

  // })
})
