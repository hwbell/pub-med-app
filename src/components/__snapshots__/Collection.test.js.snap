// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection renders correctly 1`] = `
<div
  className="outline collection"
>
  <div
    className="collapse"
    style={Object {}}
  />
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "height": "40px",
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <span
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "position": "absolute",
            "top": "0px",
            "width": "100%",
          }
        }
      >
        <p
          className="collection-title"
        >
          <strong>
            sample collection 
          </strong>
           (3)
        </p>
        <i
          className="far fa-edit"
          onClick={[Function]}
        />
      </div>
    </span>
  </div>
  <span>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": "10px",
        }
      }
    >
      <button
        aria-label={null}
        className="btn btn-link btn-md"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "16px",
          }
        }
        type="button"
      >
        <i
          className="far fa-file-pdf"
        />
      </button>
      <button
        aria-label={null}
        className="btn btn-link btn-md"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "16px",
          }
        }
        type="button"
      >
        <i
          className="far fa-save"
        />
      </button>
      <button
        aria-label={null}
        className="btn btn-link btn-md"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "16px",
          }
        }
        type="button"
      >
        <i
          className="warn-icon fas fa-trash"
        />
      </button>
    </div>
    <div
      className="results-holder"
    >
      <div
        className="article-result"
      >
        <div
          style={
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "space-around",
              "width": "100%",
            }
          }
        >
          <p
            className="article-title"
          />
          <p
            className="article-title"
          />
          <p
            className="article-title"
          >
            <strong>
                undefined
            </strong>
          </p>
        </div>
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "marginLeft": "15px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button view btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            view article
          </button>
          <button
            aria-label={null}
            className="article-button warn btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            remove
          </button>
        </div>
      </div>
      <div
        className="article-result"
      >
        <div
          style={
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "space-around",
              "width": "100%",
            }
          }
        >
          <p
            className="article-title"
          />
          <p
            className="article-title"
          />
          <p
            className="article-title"
          >
            <strong>
                undefined
            </strong>
          </p>
        </div>
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "marginLeft": "15px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button view btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            view article
          </button>
          <button
            aria-label={null}
            className="article-button warn btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            remove
          </button>
        </div>
      </div>
      <div
        className="article-result"
      >
        <div
          style={
            Object {
              "alignItems": "flex-start",
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "space-around",
              "width": "100%",
            }
          }
        >
          <p
            className="article-title"
          />
          <p
            className="article-title"
          />
          <p
            className="article-title"
          >
            <strong>
                undefined
            </strong>
          </p>
        </div>
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "marginLeft": "15px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button view btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            view article
          </button>
          <button
            aria-label={null}
            className="article-button warn btn btn-secondary btn-sm"
            onClick={[Function]}
            type="button"
          >
            remove
          </button>
        </div>
      </div>
    </div>
  </span>
</div>
`;
